### PR TITLE
[Swift 2.2.1] [SILGen] Fix for 'try?' leaking the error value

### PR DESCRIPTION
**Pull request by Reflejo**
#### What's in this pull request?

Backport the fix for `try?` memory leakage to Swift 2.2. This was the original PR to master: https://github.com/apple/swift/pull/1740. @tkremenek  please consider merging it for a 2.2.x release.
#### Resolved bug number: ([SR-972](https://bugs.swift.org/browse/SR-972))

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by tkremenek** [comment]
@swift-ci test

File: test/Interpreter/errors.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      expectEqual(c, LifetimeTracked(0))
+    } else {
+      expectUnreachable()
+    }
+  }
+}
+
+ErrorHandlingTests.test("tryOptional") {
+  expectEqual(LifetimeTracked(1), try? furball(false))
+  expectEqual(Optional<LifetimeTracked>.none, try? furball(true))
```
Comment (dduan) [inline test/Interpreter/errors.swift]: This test is written for Swift 3. I believe changing it to `.None` will make the test pass.

Comment (Reflejo) [inline test/Interpreter/errors.swift]: Yup, missed that one.

**Comment by Reflejo** [comment]
@tkremenek Now tests should be passing, sorry for the noise

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by Reflejo** [comment]
Test passed, I had to rebase after: https://github.com/apple/swift/pull/1798 though.

**Comment by tkremenek** [comment]
@swift-ci test and merge

**Comment by dduan** [comment]
Apparently `test and merge` doesn't work on protected branch like this one.

cc @shahmishal

**Comment by tkremenek** [comment]
@swift-ci test

**Comment by tkremenek** [comment]
Makes sense for `test and merge` not to work.  Not a problem.  I'll merge directly.

**Pull request merged**